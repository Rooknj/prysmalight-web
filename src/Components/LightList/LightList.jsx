import React from "react";
import PropTypes from "prop-types";
import LightMutationContainer from "./Light/LightMutationContainer";
import Grid from "@material-ui/core/Grid";

const propTypes = {
  lights: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      connected: PropTypes.number,
      state: PropTypes.string,
      brightness: PropTypes.number,
      color: PropTypes.shape({
        r: PropTypes.number.isRequired,
        g: PropTypes.number.isRequired,
        b: PropTypes.number.isRequired
      }),
      effect: PropTypes.string,
      speed: PropTypes.number,
      supportedEffects: PropTypes.array
    })
  )
};

const defaultProps = {
  lights: []
};

const LightList = ({ lights }) => (
  <Grid container justify={lights.length < 4 ? "space-around" : "flex-start"}>
    {lights.map(light => (
      <Grid key={light.id} item xs={12} sm={12} md={6} lg={4} xl={3}>
        <LightMutationContainer light={light} />
      </Grid>
    ))}
  </Grid>
);

LightList.propTypes = propTypes;
LightList.defaultProps = defaultProps;

export default LightList;
