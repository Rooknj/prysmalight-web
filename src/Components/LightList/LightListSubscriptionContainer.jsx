import React from "react";
import PropTypes from "prop-types";
import {
  LIGHTS_CHANGED,
  LIGHT_ADDED,
  LIGHT_REMOVED
} from "../graphqlConstants";
import LightList from "./LightList";

const propTypes = {
  lights: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      connected: PropTypes.number,
      state: PropTypes.string,
      brightness: PropTypes.number,
      color: PropTypes.shape({
        r: PropTypes.number.isRequired,
        g: PropTypes.number.isRequired,
        b: PropTypes.number.isRequired
      }),
      effect: PropTypes.string,
      speed: PropTypes.number,
      supportedEffects: PropTypes.array
    })
  ),
  subscribeToLightChanges: PropTypes.func
};

const defaultProps = {
  subscribeToLightChanges: () => {}
};

const addLightToQuery = (cacheData, { subscriptionData }) => {
  // If no data was returned, do nothing
  if (!subscriptionData) return cacheData;
  // Find the id of the new light
  const newLight = subscriptionData.data.lightAdded;
  // If the light already exists in the cache, do nothing
  if (cacheData.lights.find(light => light.id === newLight.id)) {
    return cacheData;
  }
  // If the light doesnt exist, add it to the cache
  return Object.assign({}, cacheData, {
    lights: [...cacheData.lights, newLight]
  });
};

const removeLightFromQuery = (cacheData, { subscriptionData }) => {
  // If no data was returned, do nothing
  if (!subscriptionData) return cacheData;

  // Find the index of the light to be removed and remove it
  const lightToRemove = subscriptionData.data.lightRemoved;
  return Object.assign({}, cacheData, {
    lights: cacheData.lights.filter(light => light.id !== lightToRemove)
  });
};

class LightListSubscriptionContainer extends React.Component {
  componentDidMount() {
    this.props.subscribeToLightChanges({
      document: LIGHTS_CHANGED
    });
    this.props.subscribeToLightChanges({
      document: LIGHT_ADDED,
      updateQuery: addLightToQuery
    });
    this.props.subscribeToLightChanges({
      document: LIGHT_REMOVED,
      updateQuery: removeLightFromQuery
    });
  }

  render() {
    const { lights } = this.props;
    return <LightList lights={lights} />;
  }
}

LightListSubscriptionContainer.propTypes = propTypes;
LightListSubscriptionContainer.defaultProps = defaultProps;

export default LightListSubscriptionContainer;
