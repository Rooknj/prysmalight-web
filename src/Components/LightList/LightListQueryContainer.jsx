import React from "react";
//import PropTypes from "prop-types";
import { Query } from "react-apollo";
import { GET_LIGHTS } from "../graphqlConstants";
import LightListSubscriptionContainer from "./LightListSubscriptionContainer";
import ErrorPage from "./ErrorPage";
import LinearProgress from "@material-ui/core/LinearProgress";
import Typography from "@material-ui/core/Typography";

const propTypes = {};

const defaultProps = {};

export const Loading = () => (
  <React.Fragment>
    <LinearProgress color="secondary" variant="query" />
    <br />
    <center>
      <Typography variant="h3" color="default">
        Loading...
      </Typography>
      {/* TODO: Add a refresh button here */}
    </center>
  </React.Fragment>
);

const LightListQueryContainer = () => (
  <Query query={GET_LIGHTS} fetchPolicy="cache-and-network" notifyOnNetworkStatusChange>
    {({ loading, error, data, subscribeToMore, refetch, networkStatus }) => {
      if (loading || networkStatus === 4) return <Loading />;
      if (error) return <ErrorPage message={error.message} refetch={refetch} />;
      return (
        <LightListSubscriptionContainer
          lights={data.lights}
          subscribeToLightChanges={subscribeToMore}
        />
      );
    }}
  </Query>
);

LightListQueryContainer.propTypes = propTypes;
LightListQueryContainer.defaultProps = defaultProps;

export default LightListQueryContainer;
