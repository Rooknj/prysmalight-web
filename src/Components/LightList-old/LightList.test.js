import React from "react";
import { shallow, mount } from "enzyme";
import LightList from "./LightList";
import { createSerializer } from "enzyme-to-json";

// Mock out container components so that they will not get rendered while mounting
jest.mock("./Light/LightMutationContainer", () => () => (
  <div id="MockLightMutationContainer">Mocked Light Mutation Container</div>
));

// Add serializer to allow enzyme to create snapshots
expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

// TODO: Test rendering with actual mock lights
it("renders without crashing", () => {
  shallow(<LightList />);
});

it("matches shallow snapshot", () => {
  const shallowed = shallow(<LightList />);
  expect(shallowed).toMatchSnapshot();
});

it("matches mounted snapshot", () => {
  const mounted = mount(<LightList />);
  expect(mounted).toMatchSnapshot();
});
