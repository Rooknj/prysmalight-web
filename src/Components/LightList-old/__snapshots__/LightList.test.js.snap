// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches mounted snapshot 1`] = `
<div
  className="MuiGrid-container-1 MuiGrid-justify-xs-space-around-21"
/>
`;

exports[`matches shallow snapshot 1`] = `
<WithStyles(Grid)
  container={true}
  justify="space-around"
/>
`;
